CRM/Core/BAO/SchemaHandler.php

```diff
@@ -658,4 +686,68 @@ public static function addIndexSignature($table, &$indices) {
     }
   }
 
+  /**
+   * Compare the indices specified in the XML files with those in the DB.
+   *
+   * @return array
+   *   index specifications
+   */
+  public static function getMissingIndices() {
+    $requiredSigs = $existingSigs = array();
+    // Get the indices defined (originally) in the xml files
+    $requiredIndices = CRM_Core_DAO_AllCoreTables::indices();
+    foreach ($requiredIndices as $table => $indices) {
+      $reqSigs[] = CRM_Utils_Array::collect('sig', $indices);
+    }
+    CRM_Utils_Array::flatten($reqSigs, $requiredSigs);
+
+    // Get the indices in the database
+    $existingIndices = CRM_Core_BAO_SchemaHandler::getIndexes(array_keys($requiredIndices));
+    foreach ($existingIndices as $table => $indices) {
+      CRM_Core_BAO_SchemaHandler::addIndexSignature($table, $indices);
+      $extSigs[] = CRM_Utils_Array::collect('sig', $indices);
+    }
+    CRM_Utils_Array::flatten($extSigs, $existingSigs);
+
+    // Compare
+    $missingSigs = array_diff($requiredSigs, $existingSigs);
+    // Get missing indices
+    $missingIndices = array();
+    foreach ($missingSigs as $sig) {
+      $sigParts = explode('::', $sig);
+      foreach ($requiredIndices[$sigParts[0]] as $index) {
+        if ($index['sig'] == $sig) {
+          $missingIndices[$sigParts[0]][] = $index;
+          continue;
+        }
+      }
+    }
+    return $missingIndices;
+  }
+
+  /**
+   * Create missing indices.
+   *
+   * @param array $missingIndices as returned by getMissingIndices()
+   */
+  public static function createMissingIndices($missingIndices) {
+    $queries = array();
+    foreach ($missingIndices as $table => $indexList) {
+      foreach ($indexList as $index) {
+        $queries[] = "CREATE " .
+        (array_key_exists('unique', $index) && $index['unique'] ? 'UNIQUE ' : '') .
+        "INDEX {$index['name']} ON {$table} (" .
+          implode(", ", $index['field']) .
+        ")";
+      }
+    }
+
+    /* FIXME potential problem if index name already exists, so check before creating */
+    $dao = new CRM_Core_DAO();
```

Comment:
definitely on the nice to have - although not a must have if doing outside of upgrade. I think the existing addIndex would cope here?